nicsdru_nidirect_theme.theme

```diff
@@ -49,6 +49,8 @@ function nicsdru_nidirect_theme_preprocess_html(array &$variables) {
 function nicsdru_nidirect_theme_preprocess_page(array &$variables) {
   $layout_classes = ['grid', 'layout--main-n-side'];
   $search_layout_classes = ['grid', 'layout--side-n-main', 'search-content'];
+
+  // Process node pages.
   $node = Drupal::routeMatch()->getParameter('node');
   if (is_object($node) && !empty($node->bundle())) {
```

Comment:
You could probably simplify the check here to something like: `if ($node instanceof NodeInterface) {` which should, I think, handle empty/null values as well as confirm whether it's the correct entity type.

Reply:
Tried `if ($node instanceof NodeInterface) {` - but it seems $node is not an instance of NodeInterface so stopped the code working.

`kint(get_class($node))` gives "Drupal\node\Entity\Node" - so changed to:

`if ($node instanceof Drupal\node\Entity\Node) {`

Reply:
If you add 'use Drupal\node\Entity\NodeInterface' at the top of the theme file you could then use $node instanceof NodeInterface. It's good practice to code against interfaces rather than implementations.

Reply:
@omahm  @johangant - tried adding 'use Drupal\node\Entity\NodeInterface' but no joy.  $node is defo not an instance of NodeInterface and so `if ($node instanceof NodeInterface) {` resolves to FALSE.

These seem relevant: 
https://www.drupal.org/project/drupal/issues/3077036
https://www.drupal.org/project/drupal/issues/2730631

Based on those, `\Drupal::routeMatch()->getParameter('node')` should return a node object - but sometimes it returns a node id on certain routes.

So on that basis `if (is_object($node) && !empty($node->bundle())) {` maybe actually makes sense? 

This also works - `if ($node instanceof Node) {` ...

Confused 😕

Reply:
Ahh I know what that is, that's the route type param. You can have a nid parameter and have it resolve to a node or just keep as an integer.

Reply:
I would first check if it's an instance of NodeInterface and if that returns false then pass the value into entitytypemanager to load the node and then continue on from there with your bundle call

Reply:
@omahm @johangant 

OK - so thinking this is getting more complex than it should be.

So this is a template_preprocess_page() hook, and the node is already loaded and made available in $variables['node'](see https://api.drupal.org/api/drupal/core%21includes%21theme.inc/function/template_preprocess_page/9.0.x).

Incidentally the code used to populate $variables['node'] is this:
```
  if ($node = \Drupal::routeMatch()
    ->getParameter('node')) {
    $variables['node'] = $node;
  }
```

This is what I have in the latest commit:
```
  // Process node pages.
  $node = $variables['node'];
  if (isset($node) && method_exists($node, 'bundle')) {
    switch ($node->bundle()) {
      case "article":
      case "application":
      case "gp_practice":
      case "health_condition":
      case "news":
      case "page":
        $variables['main_content_attr']->addClass($layout_classes);
        $variables['sidebar_second_attr']->addClass('ga-aside');
        break;

      case "contact":
        $variables['main_content_attr']->addClass($layout_classes);
        $variables['sidebar_second_attr']->addClass('ga-aside');
        break;
    }
  }
```

Does this seem OK or am I missing something?

Reply:
That's ok but it won't process any routes that pass the node parameter by integer.Noticed the namespace for NodeInterface is wrong in the previous comments which is why is didn't work.

`
use \Drupal\node\NodeInterface;

// Top of the theme code..... 

`if (! $node instanceof NodeInterface) {
    $node = \Drupal::entityTypeManager()->getStorage('node')->load($node);
  }

// continue on with the rest of the code....
if (isset($node) && method_exists($node, 'bundle')) 
'

Basically this code says "If we don't have a Node entity the node parameter is possibly an integer, so load the node with that ID.

Reply:
Cheers Steve.

From reading around this, I don't think this approach for loading nodes is something a themer would normally use and its not documented at all - https://www.drupal.org/docs/8/theming-drupal-8/modifying-attributes-in-a-theme-file.

Training courses on Drupalize.me covering the .theme file and preprocess functions never get into this kind of entity API approach for loading nodes - you are shown/expected to work with the $variables passed into the preprocess function. 

I think there might be times you need to do more to get at the node in a preprocess function, and maybe use the nodeinterface / entitymanager code you've shown above.

But I can't help feeling that $variables['node'] should be the first port of call - and it should be the node object.  Routes and route controllers should be doing this "upcasting" thing ensure the node id is upcast to the full object for the purposes of theming preprocess functions - especially theme_preprocess_page.

I'm keen to get this right so that there is an established approach that any themer could work with - whether they are noobs like me or more experienced Drupal themers ...